### DISCUSSION TITLE
Xbox One/Series Port

### DISCUSSION THREAD

**Discussion by fwannmacher** [diasurgical/DevilutionX#3599] [Ideas]
A UWP Xbox port would be awesome :)

**Comment by AJenbo** [comment]
Ok, do you have an Xbox One and are you willing to work on porting it?

**Comment by fwannmacher** [comment]
I do have a Xbox but I don't have a Windows env to work on it sadly

**Reply by qndel** [reply]
is windows env necessary to work on xbox? (Got no clue about it)

**Reply by fwannmacher** [reply]
as far as I know it is :( https://docs.microsoft.com/en-us/windows/uwp/xbox-apps/getting-started

**Comment by fwannmacher** [comment]
It looks like SDL even supports UWP https://github.com/libsdl-org/SDL/blob/350ca0f908b5ec450a13e54e9158d35622ffc6f7/docs/README-winrt.md
If someone here wants to port to UWP then I can test it on my Xbox

**Comment by fwannmacher** [comment]
I've been working on this and I got it running on Xbox but without sound. As soon as I get it fixed I'll make PR for discussing possible improvements

**Reply by AJenbo** [reply]
Exciting. I'm curious to see what rout you have gone with for porting it.

**Reply by fwannmacher** [reply]
Basically I have used the Visual Studio compile process to generate a static lib instead of an executable and then I have set up a UWP project using the static lib and a compiled WinRT SDL lib. I'm not sure what to do with the SDL lib though. I was thinking about just adding it to the repo alongside the VS project.

**Reply by fwannmacher** [reply]
It looks like vcpkg already handles it :) I'm not used with it as I'm not a windows dev xD

**Reply by AJenbo** [reply]
Oh, that sounds kinda similar to what we do for Android. Maybe the process can be somewhat shared between them. Except for the Java vs UWP part of cause.

**Reply by fwannmacher** [reply]
Dumb question: how do I set a specific platform using CMAKE + vcpkg? There is sdl2:x64-uwp but I have no idea how to say to vcpkg to use that specific version.

**Reply by StephenCWills** [reply]
Seems like it's determined for CMake by the `VCPKG_TARGET_TRIPLET` variable.
https://vcpkg.readthedocs.io/en/latest/users/integration/#triplet-selection

So assuming you've set up a separate build target in `CMakeSettings.json` for your Xbox build process, you'd need to add `-DVCPKG_TARGET_TRIPLET=x64-uwp` in the `cmakeCommandArgs` field (also labeled "CMake command arguments" in the Visual Studio UI).

FYI, as far as I can tell, this will dictate the triplet for all your dependencies, and there's no way to override the triplet specifically for one library. I imagine it doesn't make a lot of sense to mix and match triplets since you'd essentially be selecting different target compilers/runtimes.

Also FYI, if you call it a dumb question, you'll make me feel dumb too for having to look it up. 😛

**Reply by fwannmacher** [reply]
Thanks dude!

**Comment by fwannmacher** [comment]
Are there restrictions related to art when doing splash screen, icons, etc?

**Reply by AJenbo** [reply]
No nudity :p

On 3DS there is the startup sound from the shareware installer. We have also made our own icons in the style of the originals that you can find in https://github.com/diasurgical/devilutionX/tree/master/Packaging/resources
For splash screen on iOS we just do a black screen since the build in splash screen fade from black.

Does that cover what you had in mind?

**Reply by fwannmacher** [reply]
Sure! I just didn't want to do something wrong :)

**Comment by fwannmacher** [comment]
I finally got sound working! You have no idea about the time I have lost trying to understand why even setting things to WinRT on SDL I was still getting a dependency on `winmm` which is not compatible with UWP. Then I found this... https://github.com/diasurgical/devilutionX/blob/635f2f88743d33a9d46922a231e9f9a5ef41a471/Source/utils/display.cpp#L198
:/

**Comment by fwannmacher** [comment]
I'm trying to remap the buttons for making them to work in a more Xbox-ish way but a weird thing is happening: it is like 2 buttons are pressed at the same time. Have you seen that before? I'm trying to set the new values through the cmake file.

**Reply by AJenbo** [reply]
I would suggest waiting with that, we plan to implement multiple mapping scheme as well as custome ones so that would come automatically down the line in any case.

**Reply by fwannmacher** [reply]
anyways i think i found the issue. as i'm using SDL2 then GameController is used and it doesn't have a way to replace its values like the SDL1 Joystick

**Reply by glebm** [reply]
SDL2 GameController buttons can be remapped in the config (https://github.com/diasurgical/devilutionX/blob/0287faeb13f7c5e15b58ea3cea560cd081d4bd85/Source/options.cpp#L347) or via the `SDL_GAMECONTROLLERCONFIG` environment variable

**Reply by fwannmacher** [reply]
@AJenbo any ETA for for those mappings? Is there an PR?

**Reply by AJenbo** [reply]
2022 😅 We have done most of the work for keyboard/mouse, which will be part of the next release. Gamepad remapping will need some refactoring that I think we won't be starting till after the release.

**Reply by fwannmacher** [reply]
I would suggest separating actions from buttons. Like we have start and select for bringing the panels and the spell bindings and select + start for the menu. I would create actions for each of those behaviors and then add the option for any button/combination for bringing them up. For example I want to map the current behaviors of select and start to L2 and R2 and then the menu to select. With the current implementation it would be a pain in the ass xD

**Reply by fwannmacher** [reply]
Other than that having separate actions for confirm and cancel. This way exiting a menu/panel and select options/itens are not related to action buttons.

**Comment by fwannmacher** [comment]
https://github.com/diasurgical/devilutionX/pull/3946

**Reply by fwannmacher** [reply]
still missing documentation but you can start looking into it ;)

**Comment by fwannmacher** [comment]
Just a visual improvement 😉 https://github.com/diasurgical/devilutionX/pull/3963
